### PR TITLE
exec: improved logging, added error checks for header and tls timeout;

### PR DESCRIPTION
**Pull request by j4k4**
- improved logging to detect errors caused by reaching the configured max elapsed time threshold
- added error check for Client.Timeout exceeded while awaiting headers
- added error check for net/http: TLS handshake timeout

### CODE REVIEW

File: pkg/exec/test_http_client.go [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+func NewTestHttpClient(timeout time.Duration, trips Trips) http.Client {
+	return http.Client{
+		Timeout:   timeout,
+		Transport: NewTestRoundTripper(trips...),
+	}
+}
+
+type Trips []trip
+
+type trip struct {
```
Comment (ajscholl) [inline pkg/exec/test_http_client.go:18]: I think I would export the type as it is part of the public interface.

File: pkg/exec/error.go [hunk 2]
```diff
@@ -1,92 +1,62 @@
...
-
-		return len(multiErr.Errors) > 0
-	}
+func (e MaxElapsedTimeError) Unwrap() error {
+	return e.err
+}
 
-	return false
+func IsMaxElapsedTimeError(err error) bool {
+	return errors.As(err, &MaxElapsedTimeError{})
```
[the remark this answers is not in the document]
Comment (j4k4) [reply pkg/exec/error.go:51]: Don't think so, as `errors.Is` is comparing values and we're interested in knowing if it's the same type.

File: pkg/exec/error_canceled.go [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+
+const RequestCanceledError = requestCanceledError("RequestCanceled")
+
+type requestCanceledError string
+
+func (e requestCanceledError) Error() string {
+	return string(e)
+}
+
+type ErrorChecker func(result interface{}, err error) ErrorType
```
[the remark this answers is not in the document]
Comment (j4k4) [reply pkg/exec/error_canceled.go:17]: copy & paste error :D

File: pkg/exec/test_http_client.go (hunk 1 above)
Comment (j4k4) [reply pkg/exec/test_http_client.go]: changed the naming a bit

File: pkg/exec/error_test.go [hunk 4]
```diff
@@ -1,75 +1,20 @@
...
-	} {
-		test := test
-		t.Run(name, func(t *testing.T) {
-			assert.Equal(t, test.isCanceled, exec.IsRequestCanceled(test.err))
-		})
-	}
+func TestIsMaxElapsedTimeError(t *testing.T) {
+	err := exec.NewMaxElapsedTimeError(time.Second, time.Minute, errors.New("something went sideways"))
+
+	assert.True(t, exec.IsMaxElapsedTimeError(err))
```
[the remark this answers is not in the document]
Comment (j4k4) [reply pkg/exec/error_test.go:15]: can do

File: pkg/exec/error.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (ajscholl) [reply pkg/exec/error.go:53]: Ah, sure. Never mind then

File: pkg/exec/executor_backoff_test.go [hunk 5]
```diff
@@ -27,8 +30,8 @@ func (s *ExecutorBackoffTestSuite) SetupTest() {
 			Blocking:        false,
 			CancelDelay:     0,
 			InitialInterval: time.Millisecond,
-			MaxInterval:     time.Second,
-			MaxElapsedTime:  time.Second,
+			MaxInterval:     time.Millisecond * 3,
+			MaxElapsedTime:  time.Millisecond * 10,
```
Comment (ajscholl) [inline pkg/exec/executor_backoff_test.go:34]: This seems to be not enough for some tests when one has bad luck with scheduling. Maybe this needs to be decided on a test by test basis?

Comment (j4k4) [reply pkg/exec/executor_backoff_test.go]: gave it a little more space

**Review by ajscholl — APPROVED** [review]

**Pull request merged**